File: examples/e2b-to-tenki-migration/verify.mjs
/**
 * Proves the Tenki side of the E2B->Tenki migration end to end:
 * create -> run a command -> capture output -> dispose. Exits non-zero on any failure.
 * Token/project from env (CI) or ~/.config/tenki/config.yaml (local `tenki login`).
 */
import { TenkiSandbox, stdoutText } from "@tenkicloud/sandbox";
import { readFileSync } from "node:fs";
import { homedir } from "node:os";

const cfg = (key) => {
	try {
		const c = readFileSync(`${homedir()}/.config/tenki/config.yaml`, "utf8");
		return (c.match(new RegExp(`^${key}:\\s*(.+)$`, "m"))?.[1] ?? "").trim();
	} catch {
		return "";
	}
};

const authToken = process.env.TENKI_AUTH_TOKEN || process.env.TENKI_API_KEY || cfg("auth_token");
const projectId = process.env.TENKI_PROJECT_ID || cfg("current_project_id");
const workspaceId = process.env.TENKI_WORKSPACE_ID || cfg("current_workspace_id");
if (!authToken) {
	console.error("No token. Set TENKI_AUTH_TOKEN, or run `tenki login`.");
	process.exit(1);
}

const tenki = new TenkiSandbox({ authToken });
let sandbox;
try {
	// E2B: await Sandbox.create()  ->  Tenki: createAndWait() boots + waits for ready.
	sandbox = await tenki.createAndWait({ cpuCores: 1, memoryMb: 1024, projectId, workspaceId });
	if (!sandbox.id) throw new Error("no sandbox id returned from createAndWait");

	// E2B: sandbox.commands.run("...") runs via a shell. Tenki: exec() runs a bare binary
	// + args, so port a shell command by wrapping it in `sh -c`.
	const shell = await sandbox.exec("sh", { args: ["-c", "echo hello-from-tenki"] });
	// E2B: result.stdout is a string. Tenki: bytes -> decode with stdoutText().
	const out = stdoutText(shell).trim();
	if (shell.exitCode !== 0 || out !== "hello-from-tenki") {
		throw new Error(`shell exec: exit ${shell.exitCode}, stdout ${JSON.stringify(out)}`);
	}

	// A bare binary + args needs no shell (the direct exec form).
	const py = await sandbox.exec("python3", { args: ["-c", "print(6 * 7)"] });
	if (py.exitCode !== 0 || stdoutText(py).trim() !== "42") {
		throw new Error(`python exec: exit ${py.exitCode}, stdout ${JSON.stringify(stdoutText(py).trim())}`);
	}

	console.log(`✓ e2b-to-tenki-migration: create (${sandbox.id.slice(0, 8)}…) → exec sh + python3 → "${out}" / 42 → dispose`);
} catch (e) {
	console.error("✗ " + (e?.message ?? e));
	process.exitCode = 1;
} finally {
	if (sandbox) {
		// E2B: await sandbox.kill()  ->  Tenki: close() terminates the sandbox.
		try {
			await sandbox.close();
		} catch {
			/* self-reaps via idle/lifetime caps */
		}
	}
}

File: examples/modal-to-tenki-migration/verify.py
"""
Proves the Tenki side of the Modal->Tenki migration end to end: create -> exec a
bare binary -> exec a shell one-liner -> a file round-trip -> dispose. Asserts each
result and exits non-zero on any failure. No LLM needed — this is the whole example,
asserted.

Needs Python 3.10+ and the dep in requirements.txt. Token/project from env (CI) or
~/.config/tenki/config.yaml (local `tenki login`).
"""
import os
import sys

from tenki_sandbox import Sandbox


def cfg(key):
    try:
        with open(os.path.expanduser("~/.config/tenki/config.yaml")) as f:
            for line in f:
                if line.startswith(key + ":"):
                    return line.split(":", 1)[1].strip()
    except FileNotFoundError:
        pass
    return ""


token = os.environ.get("TENKI_AUTH_TOKEN") or os.environ.get("TENKI_API_KEY") or cfg("auth_token")
if not token:
    print("No token. Set TENKI_AUTH_TOKEN, or run `tenki login`.")
    sys.exit(1)

# Python SDK auth gap: a bare `tenki login` session token must be sent as a cookie
# (prefix `cookie:`); a `tk_` API key works as-is. (The Node SDK auto-detects.)
if not token.startswith(("tk_", "ory_st_", "cookie:")):
    token = f"cookie:{token}"

opts = {"auth_token": token, "cpu_cores": 1, "memory_mb": 1024}
project_id = os.environ.get("TENKI_PROJECT_ID") or cfg("current_project_id")
workspace_id = os.environ.get("TENKI_WORKSPACE_ID") or cfg("current_workspace_id")
if project_id:
    opts["project_id"] = project_id
if workspace_id:
    opts["workspace_id"] = workspace_id

try:
    # Modal: Sandbox.create(app=app) + terminate(). Tenki: no App object; the `with`
    # block disposes the microVM at scope end.
    with Sandbox.create(**opts) as sb:
        if not sb.id:
            raise AssertionError("no sandbox id returned from create()")

        # Modal's exec() returns a live process (drain p.stdout.read(), p.wait()).
        # Tenki's exec() blocks and returns a finished CommandResult.
        r = sb.exec("python3", "-c", "print(6 * 7)")
        if not (r.exit_code == 0 and r.stdout_text.strip() == "42"):
            raise AssertionError(f"python3 exec: exit {r.exit_code}, stdout {r.stdout_text!r}")

        # Both run a bare binary + args (no shell). For shell syntax, ask for a shell
        # explicitly on either platform — same as Modal.
        sh = sb.exec("sh", "-c", "echo hello-from-tenki")
        if not (sh.exit_code == 0 and sh.stdout_text.strip() == "hello-from-tenki"):
            raise AssertionError(f"sh exec: exit {sh.exit_code}, stdout {sh.stdout_text!r}")

        # File round-trip. Modal: `with sb.open(path, "w") as f: f.write(...)`.
        # Tenki: `sb.fs.write_text(path, ...)` / `sb.fs.read_text(path)`. Tenki
        # confines file I/O to /home/tenki, so use a relative path.
        sb.fs.write_text("migrate.txt", "modal->tenki")
        back = sb.fs.read_text("migrate.txt")
        if back != "modal->tenki":
            raise AssertionError(f"fs round-trip: read back {back!r}")

    print(
        f"✓ modal-to-tenki-migration: create ({sb.id[:8]}…) → "
        f"exec python3 + sh + fs round-trip → 42 → dispose"
    )
except Exception as e:  # noqa: BLE001
    print(f"✗ {type(e).__name__}: {e}")
    sys.exit(1)
